File: api/smartdl/audius_client.py
import logging
import httpx

logger = logging.getLogger("smartdl.audius")

AUDIUS_API = "https://api.audius.co"

AUDIUS_TIMEOUT = httpx.Timeout(15.0, connect=8.0)


async def _audius_get(path: str) -> dict | None:
    try:
        async with httpx.AsyncClient(timeout=AUDIUS_TIMEOUT) as client:
            resp = await client.get(
                f"{AUDIUS_API}{path}",
                headers={"User-Agent": "Sinc/1.0"},
            )
            if resp.status_code == 200:
                return resp.json()
    except Exception as e:
        logger.warning("Audius request failed: %s", e)
    return None


async def audius_search(query: str, limit: int = 5) -> list[dict]:
    data = await _audius_get(f"/v1/tracks/search?query={httpx.escape(query)}&limit={limit}")
    if not data or "data" not in data:
        return []
    results = []
    for t in data["data"]:
        user = t.get("user", {}) or {}
        results.append({
            "id": t.get("id"),
            "title": t.get("title", "Unknown"),
            "uploader": user.get("name") or user.get("handle", "Unknown"),
            "url": t.get("id"),
            "duration": t.get("duration", 0),
            "downloadable": t.get("downloadable", False),
            "stream_url": f"{AUDIUS_API}/v1/tracks/{t['id']}/stream" if t.get("id") else None,
        })
    return results


async def audius_get_audio_url(artist: str, title: str) -> str | None:
    query = f"{artist} - {title}"
    results = await audius_search(query, limit=3)
    for r in results:
        stream = r.get("stream_url")
        if stream:
            logger.info("audius: found stream for %s - %s", artist, title)
            return stream
        if r.get("downloadable") and r.get("id"):
            dl_url = f"{AUDIUS_API}/v1/tracks/{r['id']}/download"
            logger.info("audius: found download for %s - %s", artist, title)
            return dl_url
    return None

File: app/src/main/kotlin/com/sinc/enhanced/data/remote/AudiusClient.kt
package com.sinc.enhanced.data.remote

import android.util.Log
import kotlin.jvm.Synchronized
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.net.URLEncoder

/** TODO: Convert all public methods to suspend functions */
class AudiusClient(private val client: OkHttpClient) {

    data class AudiusTrack(
        val id: String,
        val title: String,
        val artist: String,
        val duration: Long,
        val artworkUrl: String?,
        val streamUrl: String?,
        val genre: String? = null
    )

    @Volatile private var cachedNode: String? = null
    @Volatile private var lastNodeCheck: Long = 0

    @Synchronized
    private fun getNode(): String {
        cachedNode?.let {
            if (System.currentTimeMillis() - lastNodeCheck < 3600000L) return it
        }
        val defaultNode = "https://discoveryprovider.audius.co"
        return try {
            val request = Request.Builder()
                .url("https://api.audius.co")
                .build()
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) return@use defaultNode
                val json = JSONObject(response.body?.string() ?: return@use defaultNode)
                val nodes = json.optJSONArray("data") ?: return@use defaultNode
                val node = nodes.optString(0, "").trimEnd('/')
                if (node.isNotEmpty()) {
                    cachedNode = node
                    lastNodeCheck = System.currentTimeMillis()
                    node
                } else defaultNode
            }
        } catch (e: Exception) { Log.e("AudiusClient", "getNode failed", e); defaultNode }
    }

    fun search(query: String, limit: Int = 5): List<AudiusTrack> {
        val node = getNode()
        val url = "$node/v1/full/tracks/search?query=${URLEncoder.encode(query, "UTF-8")}&limit=$limit&app_name=SincEnhanced"
        return try {
            val request = Request.Builder()
                .url(url)
                .header("Accept", "application/json")
                .build()
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) return@use emptyList()
                val json = JSONObject(response.body?.string() ?: return@use emptyList())
                val data = json.optJSONObject("data") ?: return@use emptyList()
                val tracks = data.optJSONArray("tracks") ?: return@use emptyList()
                (0 until minOf(tracks.length(), limit)).mapNotNull { i ->
                    val item = tracks.getJSONObject(i)
                    val user = item.optJSONObject("user")
                    AudiusTrack(
                        id = item.optString("track_id") ?: item.optString("id"),
                        title = item.optString("title"),
                        artist = user?.optString("name") ?: item.optString("handle") ?: "Unknown",
                        duration = item.optLong("duration") ?: 0L,
                        artworkUrl = item.optString("artwork").ifEmpty { null },
                        streamUrl = item.optString("download_url")
                            ?: "$node/v1/tracks/${item.optString("track_id") ?: item.optString("id")}/stream?app_name=SincEnhanced",
                        genre = item.optString("genre").ifEmpty { null }
                    )
                }
            }
        } catch (e: Exception) { Log.e("AudiusClient", "search failed", e); emptyList() }
    }

    fun getStreamUrl(trackId: String): String? {
        val node = getNode()
        val url = "$node/v1/tracks/$trackId/stream?app_name=SincEnhanced"
        return try {
            val request = Request.Builder().url(url).build()
            client.newCall(request).execute().use { response ->
                if (response.isSuccessful) url else null
            }
        } catch (e: Exception) { Log.e("AudiusClient", "getStreamUrl failed", e); null }
    }
}
